services/sn_analysis_service/app.py

```diff
@@ -112,74 +134,154 @@ async def run_sn_model(dd):
+            result_data = base64.b64encode(open(f.name, "rb").read())
+            local_temp_files.append(f.name)
+
+            analysis_results = {
+                "inference_data": {"format": "netcdf4", "data": inference_data},
+                "plots": [{"format": "png", "data": plot_data}],
+                "results": {"format": "joblib", "data": result_data},
+            }
+            rez.update(
+                {
+                    "analysis": analysis_results,
+                    "status": "success",
+                    "message": f"Good results with chi^2/dof={result.chisq/result.ndof}",
+                }
+            )
         else:
-            results.update({"status": "failure", "message": "model failed to converge"})
+            log("Fit failed.")
+            rez.update({"status": "failure", "message": "model failed to converge"})
 
     except Exception as e:
-        results.update(
-            {"status": "failure", "message": f"problem running the model {e}"}
+        log(f"Exception while running the model: {e}")
+        log(f"{traceback.format_exc()}")
+        log(f"Data: {data}")
+        rez.update({"status": "failure", "message": f"problem running the model {e}"})
+    finally:
+        # clean up local files
+        for f in local_temp_files:
+            try:
+                os.remove(f)
+            except:  # noqa E722
+                pass
+    return rez
+
+
+class MainHandler(tornado.web.RequestHandler):
+    def set_default_headers(self):
+        self.set_header('Content-Type', 'application/json')
+
+    def error(self, code, message):
+        self.set_status(code)
+        self.write({'message': message})
+
+    def get(self):
+        self.write({'status': 'active'})
+
+    def post(self):
+        """
+        Analysis endpoint which sends the `data_dict` off for
+        processing, returning immediately. The idea here is that
+        the analysis model may take awhile to run so we
+        need async behavior.
+        """
+        try:
+            data_dict = tornado.escape.json_decode(self.request.body)
+        except json.decoder.JSONDecodeError:
+            err = traceback.format_exc()
+            log(f"JSON decode error: {err}")
+            return self.error(400, "Invalid JSON")
+
+        required_keys = ["inputs", "callback_url", "callback_method"]
+        for key in required_keys:
+            if key not in data_dict:
+                log(f"missing required key {key} in data_dict")
+                return self.error(400, f"missing required key {key} in data_dict")
+
+        def sn_analysis_done_callback(
+            future,
+            logger=log,
+            data_dict=data_dict,
+        ):
+            """
+            Callback function for when the sn analysis service is done.
+            Sends back results/errors via the callback_url.
+            """
+            try:
+                result = future.result()
+            except Exception as e:
```

Comment:
You won't get an exception here; this doesn't do an await.

Reply:
Incorrect. This is inside a callback and future.result() will error if the coroutine didn't return something, but errored instead. This is the way to catch errors from coroutines and log them appropriately.

Reply:
See https://quantlane.com/blog/ensure-asyncio-task-exceptions-get-logged/  ("HOW TO ENSURE TASK EXCEPTIONS GET LOGGED IMMEDIATELY")